### ISSUE TITLE
Support php 8.1

### ISSUE THREAD

**Issue by jasverix** [php-mime-mail-parser/php-mime-mail-parser#380]
**Describe the bug**
Php 8.1 will throw a deprecated error if the inheritors of ArrayAccess does not include a return type. Adding return types to offsetGet, offsetSet, offsetExists and offsetUnset will make this component work in php 8.1 too.

Adding the `#[\ReturnTypeWillChange]` attribute will also solve the problem - and it even works back to php 4, since # is a comment before php 8.

**Version (please complete the following information):**
 - Version of PHP 8.1
 - Version of php-mime-mail-parser 7.0.0

**Additional context**
Fatal error: During inheritance of ArrayAccess: Uncaught PhpErrorException: Return type of PhpMimeMailParser\MimePart::offsetExists($offset) should either be compatible with ArrayAccess::offsetExists(mixed $offset): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in .../vendor/php-mime-mail-parser/php-mime-mail-parser/src/MimePart.php:95

**Comment by mbretter** [comment]
any chance getting this fixed soon?

**Comment by eXorus** [comment]
Sorry I didn't update the issue.

The release has been done 12 days ago with the compatibility of PHP8.1.

For your information the master branch is the next major version so this PR is needed but later.

**Comment by mbretter** [comment]
hmm, the issue still exists with version 7.1:
```
composer show | grep mim
php-mime-mail-parser/php-mime-mail-parser 7.1.0              A fully tested email parser for PHP 7.2+ (mailparse extension wrapper).

PHP Deprecated:  Return type of PhpMimeMailParser\MimePart::offsetExists($offset) should either be compatible with ArrayAccess::offsetExists(mixed $offset): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /data/www/sundesk/api/vendor/php-mime-mail-parser/php-mime-mail-parser/src/MimePart.php on line 95
PHP Deprecated:  Return type of PhpMimeMailParser\MimePart::offsetGet($offset) should either be compatible with ArrayAccess::offsetGet(mixed $offset): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /data/www/sundesk/api/vendor/php-mime-mail-parser/php-mime-mail-parser/src/MimePart.php on line 111
PHP Deprecated:  Return type of PhpMimeMailParser\MimePart::offsetSet($offset, $value) should either be compatible with ArrayAccess::offsetSet(mixed $offset, mixed $value): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /data/www/sundesk/api/vendor/php-mime-mail-parser/php-mime-mail-parser/src/MimePart.php on line 83
PHP Deprecated:  Return type of PhpMimeMailParser\MimePart::offsetUnset($offset) should either be compatible with ArrayAccess::offsetUnset(mixed $offset): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /data/www/sundesk/api/vendor/php-mime-mail-parser/php-mime-mail-parser/src/MimePart.php on line 103


php -v
PHP 8.1.4 (cli) (built: Apr  4 2022 13:30:17) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.4, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.4, Copyright (c), by Zend Technologies
    with Xdebug v3.1.2, Copyright (c) 2002-2021, by Derick Rethans

```

cheers

**Comment by jimmybrancaccio** [comment]
I had to run:

```
composer remove php-mime-mail-parser/php-mime-mail-parser
composer require php-mime-mail-parser/php-mime-mail-parser
```

```
php -v
PHP 8.1.4 (cli) (built: Mar 29 2022 00:29:09) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.4, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.4, Copyright (c), by Zend Technologies
```

**Comment by mbretter** [comment]
when checking MimePart.php with the 7.1 tag, there are no type hints:

https://github.com/php-mime-mail-parser/php-mime-mail-parser/blob/7.1.0/src/MimePart.php

**Comment by eXorus** [comment]
Maybe I did something wrong, I'm not an expert in git and I try to get the last release, update the code for 8.1 and do another release without impacting master branch.

In the PR the tests on PHP8.1 was ok without warnings:
https://github.com/php-mime-mail-parser/php-mime-mail-parser/pull/384

**Comment by eXorus** [comment]
Yes I didn't change this file because the tests were green, the only issue with PHP8.1 was about PHPUnit with `assertFileDoesNotExist`.

**Comment by mbretter** [comment]
Maybe you have disabled the deprecation warnings?
However the function definitions should look like this (as mentioned by @jasverix )
```php

    /**
     * ArrayAccess
     */
    #[\ReturnTypeWillChange]
    public function offsetSet($offset, $value)
    {
        if (is_null($offset)) {
            $this->part[] = $value;
            return;
        }
        $this->part[$offset] = $value;
    }

    /**
     * ArrayAccess
     *
     */
    #[\ReturnTypeWillChange]
    public function offsetExists($offset)
    {
        return isset($this->part[$offset]);
    }

    /**
     * ArrayAccess
     */
    #[\ReturnTypeWillChange]
    public function offsetUnset($offset)
    {
        unset($this->part[$offset]);
    }

    /**
     * ArrayAccess
     */
    #[\ReturnTypeWillChange]
    public function offsetGet($offset)
    {
        return isset($this->part[$offset]) ? $this->part[$offset] : null;
    }
```

**Comment by eXorus** [comment]
Could you check the last release 7.1.1? I hope it will be ok because I didn't reproduce the error I don't have PHP8.1 yet.

**Comment by mbretter** [comment]
nope, there is something completely wrong, you have declared your constructors private!?

Parser.php
```
private function __construct(ParserConfig $parserConfig = null)
    {
        $this->parserConfig = $parserConfig ?? new ParserConfig;
    }
```

which  leads to:
 Error: Call to private PhpMimeMailParser\Parser::__construct() from scope ...

MimePart.php is not present anymore, maybe you have accidentally merged a different codebase?


```
 composer show | grep mime
php-mime-mail-parser/php-mime-mail-parser 7.1.1              A fully tested email parser for PHP 7.3+ (mailparse extension wrapper).
```

**Comment by eXorus** [comment]
You're right, I deleted the last release.
I will take a look later, I did:
```
git checkout tags/7.0.0
git switch -c php81-fix
```

**Comment by mbretter** [comment]
```
git clone https://github.com/php-mime-mail-parser/php-mime-mail-parser
git checkout  php81-fix
```
this shows the "right" MimePart.php to me

switch -c creates a new branch, if I am right.
I am using checkout only, btw

**Comment by mbretter** [comment]
when using:
```
"php-mime-mail-parser/php-mime-mail-parser": "dev-php81-fix"
```
everything is working, MimePart. contains the required typehints

**Comment by eXorus** [comment]
```
git checkout tags/7.1.0
git switch -c 7.1.2
```

I don't really know where I mess up last time but I think it's good now.

[Release 7.1.2](https://github.com/php-mime-mail-parser/php-mime-mail-parser/releases/tag/7.1.2) fix the deprecated error on PHP 8.1 with the new attribute.

Let me know if there is another issue.

**Comment by mbretter** [comment]
It is working now

thx,
cheers
